fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/NegateFilterShuttle.java

```diff
@@ -0,0 +1,93 @@
+
+import com.starrocks.analysis.BinaryType;
+import com.starrocks.sql.optimizer.operator.scalar.BinaryPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.CompoundPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.CompoundPredicateOperator.CompoundType;
+import com.starrocks.sql.optimizer.operator.scalar.InPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.IsNullPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.rewrite.BaseScalarOperatorShuttle;
+
+public class NegateFilterShuttle extends BaseScalarOperatorShuttle {
+
+    private static NegateFilterShuttle INSTANCE = new NegateFilterShuttle();
+
+    public static NegateFilterShuttle getInstance() {
+        return INSTANCE;
+    }
+
+    public ScalarOperator negateFilter(ScalarOperator scalarOperator) {
+        return scalarOperator.accept(this, null);
+    }
+
+    @Override
+    public ScalarOperator visitCompoundPredicate(CompoundPredicateOperator predicate, Void context) {
+        ScalarOperator negation;
+        if (CompoundType.NOT == predicate.getCompoundType()) {
+            negation = predicate.getChild(0);
+            if (predicate.getChild(0).isNullable()) {
+                return new CompoundPredicateOperator(CompoundType.OR, negation,
+                        new IsNullPredicateOperator(predicate.getChild(0)));
+            } else {
+                return negation;
+            }
+
+        } else {
+            negation = new CompoundPredicateOperator(CompoundType.NOT, predicate);
+            if (predicate.isNullable()) {
+                return new CompoundPredicateOperator(CompoundType.OR, negation, new IsNullPredicateOperator(predicate));
+            } else {
+                return negation;
+            }
+        }
+    }
+
+    @Override
+    public ScalarOperator visitBinaryPredicate(BinaryPredicateOperator predicate, Void context) {
+        ScalarOperator negation;
+        if (BinaryType.EQ_FOR_NULL == predicate.getBinaryType()) {
+            negation = new CompoundPredicateOperator(CompoundType.NOT, predicate);
+        } else {
+            negation = predicate.negative();
+        }
+
+        if (predicate.getChild(0).isNullable()) {
+            ScalarOperator isNull = new IsNullPredicateOperator(predicate.getChild(0));
+            return new CompoundPredicateOperator(CompoundType.OR, negation, isNull);
+        } else {
+            return negation;
+        }
+    }
+
+    @Override
+    public ScalarOperator visitInPredicate(InPredicateOperator predicate, Void context) {
+        ScalarOperator negation = new InPredicateOperator(!predicate.isNotIn(), predicate.getChildren());
+        if (predicate.getChild(0).isNullable()) {
+            ScalarOperator isNull = new IsNullPredicateOperator(predicate.getChild(0));
+            return new CompoundPredicateOperator(CompoundType.OR, negation, isNull);
```

Comment:
Outermost  "p is null" in "p or p is null" can be reduced to p, have you reduce this?

Reply:
Yes, have done. When we negate `col in (xx, xx)`, we derive the  `col not in (xx, xx)` when col is a not nullable col.

Reply:
P: (p(col) or col is null) and q(col)

if when element in col is null,  both p(col) and q(col) is false, then P is false, p(col) and q(col) is false too,
if when element in col is not null, then P can be reduced p(col) and q(col),

so P can be always reduced to p(col) and q(col) if col=false implies p(col) and q(col) is false